### PR TITLE
add Extended Stats Bucket

### PR DESCRIPTION
**Pull request by nicolassing**

### CODE REVIEW

File: lib/Elastica/Aggregation/ExtendedStatsBucket.php [hunk 1]
```diff
@@ -0,0 +1,74 @@
+<?php
+
+namespace Elastica\Aggregation;
+
+use Elastica\Exception\InvalidException;
+
+/**
+ * Class ExtendedStatsBucket.
```
Comment (thePanz) [inline lib/Elastica/Aggregation/ExtendedStatsBucket.php]: That does not add more info about the class :)
Can you add a more descriptive text here?

File: lib/Elastica/Aggregation/ExtendedStatsBucket.php [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+
+        if (null !== $bucketsPath) {
+            $this->setBucketsPath($bucketsPath);
+        }
+    }
+
+    /**
+     * Set the buckets_path for this aggregation.
+     *
+     * @return $this
```
Comment (thePanz) [inline lib/Elastica/Aggregation/ExtendedStatsBucket.php]: Not needed, it is implicit from the method signature

File: lib/Elastica/Aggregation/ExtendedStatsBucket.php [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+     */
+    public function setBucketsPath(string $bucketsPath): self
+    {
+        return $this->setParam('buckets_path', $bucketsPath);
+    }
+
+    /**
+     * Set the gap policy for this aggregation.
+     *
+     * @return $this
```
Comment (thePanz) [inline lib/Elastica/Aggregation/ExtendedStatsBucket.php]: Not needed, it is implicit from the method signature

File: lib/Elastica/Aggregation/ExtendedStatsBucket.php [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+     */
+    public function setGapPolicy(string $gapPolicy): self
+    {
+        return $this->setParam('gap_policy', $gapPolicy);
+    }
+
+    /**
+     * Set the format for this aggregation.
+     *
+     * @return $this
```
Comment (thePanz) [inline lib/Elastica/Aggregation/ExtendedStatsBucket.php]: Not needed, it is implicit from the method signature

File: lib/Elastica/Aggregation/ExtendedStatsBucket.php [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+     */
+    public function setFormat(string $format): self
+    {
+        return $this->setParam('format', $format);
+    }
+
+    /**
+     * Set the number of standard deviations for this aggregation.
+     *
+     * @return $this
```
Comment (thePanz) [inline lib/Elastica/Aggregation/ExtendedStatsBucket.php]: Not needed, it is implicit from the method signature

File: lib/Elastica/Aggregation/ExtendedStatsBucket.php [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+     */
+    public function setSigma(int $sigma): self
```
Comment (thePanz) [inline lib/Elastica/Aggregation/ExtendedStatsBucket.php]: Not needed, it is implicit from the method signature

File: test/Elastica/Aggregation/ExtendedStatsBucketTest.php [hunk 7]
```diff
@@ -0,0 +1,90 @@
...
+
+        $this->assertEquals($expected, $serialDiffAgg->toArray());
+    }
+
+    /**
+     * @group unit
+     */
+    public function testToArrayInvalidBucketsPath()
+    {
+        $this->expectException(\Elastica\Exception\InvalidException::class);
```
Comment (thePanz) [inline test/Elastica/Aggregation/ExtendedStatsBucketTest.php]: Can you import the exception and only use the `InvalidException::class` here?

File: test/Elastica/Aggregation/ExtendedStatsBucketTest.php [hunk 8]
```diff
@@ -0,0 +1,90 @@
...
+
+        $index->refresh();
+
+        return $index;
+    }
+
+    /**
+     * @group functional
+     */
+    public function testExtendedStatBucketAggregation()
```
Comment (thePanz) [inline test/Elastica/Aggregation/ExtendedStatsBucketTest.php]: Please add `(): void` as method return signature

File: test/Elastica/Aggregation/ExtendedStatsBucketTest.php [hunk 9]
```diff
@@ -0,0 +1,90 @@
...
+        $this->assertEquals(11000, $results['sum_of_squares']);
+        $this->assertEquals(66.66666666666667, $results['variance']);
+        $this->assertEquals(8.16496580927726, $results['std_deviation']);
+        $this->assertEquals(['upper' => 76.32993161855453, 'lower' => 43.670068381445475], $results['std_deviation_bounds']);
+    }
+
+    /**
+     * @group unit
+     */
+    public function testConstructThroughSetters()
```
Comment (thePanz) [inline test/Elastica/Aggregation/ExtendedStatsBucketTest.php]: Please add `(): void` as method return signature

File: test/Elastica/Aggregation/ExtendedStatsBucketTest.php [hunk 10]
```diff
@@ -0,0 +1,90 @@
...
+            ],
+        ];
+
+        $this->assertEquals($expected, $serialDiffAgg->toArray());
+    }
+
+    /**
+     * @group unit
+     */
+    public function testToArrayInvalidBucketsPath()
```
Comment (thePanz) [inline test/Elastica/Aggregation/ExtendedStatsBucketTest.php]: Please add `(): void` as method return signature

File: test/Elastica/Aggregation/ExtendedStatsBucketTest.php [hunk 11]
```diff
@@ -0,0 +1,90 @@
...
+use Elastica\Aggregation\ExtendedStatsBucket;
+use Elastica\Aggregation\Histogram;
+use Elastica\Aggregation\Max;
+use Elastica\Document;
+use Elastica\Index;
+use Elastica\Query;
+
+class ExtendedStatsBucketTest extends BaseAggregationTest
+{
+    protected function _getIndexForTest(): Index
```
Comment (thePanz) [inline test/Elastica/Aggregation/ExtendedStatsBucketTest.php]: Please use `private` as method visibility here

**Comment by nicolassing** [comment]
Hello @thePanz 

I removed all useless comments and annotations ;)

Thx !

File: test/Elastica/Aggregation/ExtendedStatsBucketTest.php (hunk 11 above)
Comment (thePanz) [reply test/Elastica/Aggregation/ExtendedStatsBucketTest.php]: This is still `protected` :)

File: lib/Elastica/Aggregation/ExtendedStatsBucket.php [hunk 12]
```diff
@@ -0,0 +1,52 @@
+<?php
+
+namespace Elastica\Aggregation;
+
+use Elastica\Exception\InvalidException;
+
+/**
+ * @see https://www.elastic.co/guide/en/elasticsearch/reference/current/search-aggregations-pipeline-extended-stats-bucket-aggregation.html
```
Comment (thePanz) [inline lib/Elastica/Aggregation/ExtendedStatsBucket.php]: Can you add just a title? 

`Implements a Extended Stats Bucket Aggregation` would be enough :)

**Comment by ruflin** [comment]
Seems CI is not happy :-( @thePanz Feel free to merge as soon as you are happy and CI is green.

File: tests/Aggregation/ExtendedStatsBucketTest.php [hunk 13]
```diff
@@ -0,0 +1,96 @@
...
+     */
+    public function testToArrayInvalidBucketsPath(): void
+    {
+        $this->expectException(InvalidException::class);
+
+        $serialDiffAgg = new ExtendedStatsBucket('bucket_part');
+        $serialDiffAgg->toArray();
+    }
+
+    protected function _getIndexForTest(): Index
```
Comment (thePanz) [inline tests/Aggregation/ExtendedStatsBucketTest.php]: this method should be set to `private`

File: test/Elastica/Aggregation/ExtendedStatsBucketTest.php (hunk 11 above)
Comment (thePanz) [reply test/Elastica/Aggregation/ExtendedStatsBucketTest.php]: This is till `protected`, it should be `private` instead

**Comment by nicolassing** [comment]
Hey @thePanz Everything is ok for you ? 
Thx !

File: src/Aggregation/ExtendedStatsBucket.php [hunk 14]
```diff
@@ -0,0 +1,54 @@
...
+use Elastica\Exception\InvalidException;
+
+/**
+ * Implements a Extended Stats Bucket Aggregation.
+ *
+ * @see https://www.elastic.co/guide/en/elasticsearch/reference/current/search-aggregations-pipeline-extended-stats-bucket-aggregation.html
+ */
+class ExtendedStatsBucket extends AbstractAggregation
+{
+    public function __construct(string $name, ?string $bucketsPath = null)
```
Comment (thePanz) [inline src/Aggregation/ExtendedStatsBucket.php]: I just noticed it now: why is the bucketPath not required in the constructor, but later required when building the request( `toArray()`)?

As it is required, I'd make it required (not empty) and remove the `setBucketPath()` method. What do you think?

Comment (thePanz) [reply src/Aggregation/ExtendedStatsBucket.php]: @nicolassing ^^

Comment (ruflin) [reply src/Aggregation/ExtendedStatsBucket.php]: @thePanz Perhaps the same argument as in the other PR might apply. No strong opinion which on is better but we should become consistent.

Comment (nicolassing) [reply src/Aggregation/ExtendedStatsBucket.php]: i followed the model of the StatsBucket aggregation :/

But both of you are right. We can make the $bucketsPath argument mandatory and remove the setter in the 2 classes :)

Comment (thePanz) [reply src/Aggregation/ExtendedStatsBucket.php]: @nicolassing : I'd leave the setters (and do the "not empty" check there, throwing an exception)
and make the argument required in the constructor

I'll cleanup the other aggregations @ruflin , but in another PR. I'd like not to have devs creating invalid objects to start with :) but this is a topic for a discussion, not here tho :)

Comment (ruflin) [reply src/Aggregation/ExtendedStatsBucket.php]: SGTM, lets try it out.

**Comment by ruflin** [comment]
@nicolassing Could you add a changelog entry? I think then this PR is good to go.

File: src/Aggregation/ExtendedStatsBucket.php (hunk 14 above)
Comment (thePanz) [reply src/Aggregation/ExtendedStatsBucket.php]: But first: let's have the `$bucketsPath` argument as required :)

File: src/Aggregation/ExtendedStatsBucket.php [hunk 15]
```diff
@@ -0,0 +1,54 @@
...
+    {
+        return $this->setParam('sigma', $sigma);
+    }
+
+    /**
+     * @throws InvalidException If buckets path or script is not set
+     */
+    public function toArray(): array
+    {
+        if (!$this->hasParam('buckets_path')) {
```
Comment (thePanz) [inline src/Aggregation/ExtendedStatsBucket.php]: This is not needed anymore

File: src/Aggregation/ExtendedStatsBucket.php [hunk 16]
```diff
@@ -0,0 +1,54 @@
...
+ *
+ * @see https://www.elastic.co/guide/en/elasticsearch/reference/current/search-aggregations-pipeline-extended-stats-bucket-aggregation.html
+ */
+class ExtendedStatsBucket extends AbstractAggregation
+{
+    public function __construct(string $name, string $bucketsPath)
+    {
+        parent::__construct($name);
+
+        if (null !== $bucketsPath) {
```
Comment (thePanz) [inline src/Aggregation/ExtendedStatsBucket.php]: the if condition is now always `true`

File: src/Aggregation/ExtendedStatsBucket.php (hunk 15 above)
Comment (nicolassing) [reply src/Aggregation/ExtendedStatsBucket.php]: Yes ! I was correcting it ^^

**Comment by thePanz** [comment]
Merged! thanks @nicolassing for the PR, the changes and the patience :+1:

**Pull request merged**